components/Input/NumericInput.tsx

```diff
@@ -1,55 +1,154 @@
-import React, { ChangeEvent, FC, memo, useCallback } from 'react';
+import { FC, memo, useCallback } from 'react';
 import styled, { css } from 'styled-components';
 
-import Input from './Input';
-
 type NumericInputProps = Omit<React.InputHTMLAttributes<HTMLInputElement>, 'onChange'> & {
-	value: string | number;
-	placeholder?: string;
-	onChange: (e: ChangeEvent<HTMLInputElement>, value: string) => void;
-	className?: string;
-	defaultValue?: any;
-	disabled?: boolean;
-	id?: string;
+	value: string;
+	onChange: (e: React.ChangeEvent<HTMLInputElement>, value: string) => void;
+	left?: React.ReactNode;
+	right?: React.ReactNode;
+	dataTestId?: string;
+	invalid?: boolean;
 	bold?: boolean;
+	textAlign?: string;
+	suffix?: string;
+	max?: number;
 };
 
 const INVALID_CHARS = ['-', '+', 'e'];
 
-const NumericInput: FC<NumericInputProps> = memo(({ onChange, bold, ...props }) => {
-	const handleOnChange = useCallback(
-		(e: ChangeEvent<HTMLInputElement>) => {
-			const standardizedNum = e.target.value.replace(/,/g, '.').replace(/[e+-]/gi, '');
-			if (isNaN(Number(standardizedNum))) return;
-			onChange(e, standardizedNum);
-		},
-		[onChange]
-	);
-
-	return (
-		<StyledInput
-			type="text"
-			inputMode="decimal"
-			onChange={handleOnChange}
-			onKeyDown={(e) => {
-				if (INVALID_CHARS.includes(e.key)) {
-					e.preventDefault();
+const isInvalid = (key: string) => INVALID_CHARS.includes(key);
+
+const NumericInput: FC<NumericInputProps> = memo(
+	({
+		value,
+		onChange,
+		left,
+		right,
+		dataTestId,
+		invalid,
+		bold,
+		textAlign,
+		max = 0,
+		className,
+		...props
+	}) => {
+		const handleChange = useCallback(
+			(e: React.ChangeEvent<HTMLInputElement>) => {
+				const standardizedNum = e.target.value.replace(/,/g, '.').replace(/[e+-]/gi, '');
+				// Consider switching to this:
+				// const standardizedNum = value
+				// .replace(/[^0-9.,]/g, '')
+				// .replace(/,/g, '.')
+				// .substring(0, 4);
+				// TODO: make regex only accept valid numbers, so we don't need to check again.
+				if (isNaN(Number(standardizedNum))) return;
+				const valueIsAboveMax = max !== 0 && Number(standardizedNum) > max;
+				if (!valueIsAboveMax) {
+					onChange(e, standardizedNum);
 				}
-			}}
-			$bold={bold}
-			{...props}
-		/>
-	);
-});
-
-export const StyledInput = styled(Input)<{ $bold?: boolean }>`
-	font-family: ${(props) => props.theme.fonts.mono};
+			},
+			[onChange, max]
+		);
+
+		return (
+			<InputContainer
+				$invalid={invalid}
+				$bold={bold}
+				$textAlign={textAlign}
+				$length={value.length}
+				className={className}
+			>
+				{left && <div className="left">{left}</div>}
+				<input
+					data-testid={dataTestId}
+					value={value}
+					type="text"
+					inputMode="decimal"
+					onChange={handleChange}
+					onKeyDown={(e) => {
+						if (isInvalid(e.key)) {
+							e.preventDefault();
+						}
+					}}
+					{...props}
+				/>
+				{right && <div className="right">{right}</div>}
```

Comment:
What do you think about creating new styled components rather than using `className`? I feel like it's a bit cleaner and reads better with styled components, also can keep a consistent pattern then?

Reply:
We can definitely do that. Sometimes (like in this case) though, I generally feel that it might be a little bit overkill to create a new styled component for just setting a margin, if we can do it from a parent styled component.

Reply:
As well as readability and consistency I also feel it's more robust to use components over parent controlled styling, like if the parent changed or was replaced it would not impact the children, as they are self contained styles.

Agree it can feel a bit overkill for single margins like this but I think it would be nice to have a standard convention for this sort of thing, as even though it's just a margin now maybe later on it changes to add a border or some other style, seems a small trade off to have a few potentially overkill styled components created to maintain a consistent pattern. Also if we instead use Spacer or some other generic utility div for small margin / padding styles we could avoid these anyway.

Reply:
I think that the spacer is actually the best solution for this particular case. I also understand why individual styled components are your preference. I guess I'm just a little wary after seeing how many of the inconsistencies we've managed to rack up have been as a result of their overuse.